Sources/PitchSpeller/Data Structures/FlowNetwork.swift

```diff
@@ -32,18 +39,78 @@ public struct FlowNetwork <Value: Hashable>: Hashable {
     /// reached zero in the flow-propagation process.
     ///
     /// - TODO: Add backflow to reversed edges.
-    public var residualNetwork: Graph<Value> {
+    public var residualNetwork: Graph<Node> {
         var residualNetwork = graph
         while let path = residualNetwork.shortestPath(from: source, to: sink) {
             residualNetwork.insertPath(path.map { $0 - maximumFlow(of: path) })
         }
         return residualNetwork
     }
+    
+    /// - Returns: (0) The maximum flow of the network and (1) the residual network produced after
+    /// pushing all possible flow from source to sink (while satisfying flow constraints) - with
+    /// saturated edges flipped and all weights removed.
+    var solvedForMaximumFlow: (Double, UnweightedGraph<Node>) {
+        var totalFlow = 0.0
+        var residualNetwork = directedGraph
+        
+        func findAugmentingPath () -> Bool {
+            guard let path = residualNetwork.shortestUnweightedPath(from: source, to: sink) else {
+                return false
+            }
+            pushFlow(through: path)
+            return true
+        }
+            
+        func pushFlow (through path: UnweightedGraph<Node>.Path) {
+            let minimumEdge = (path.adjacents.compactMap {
+                residualNetwork.weight($0)
+                }.min())!
+            totalFlow += minimumEdge
+            path.adjacents.forEach {
+                residualNetwork.updateEdge($0, with: {
+                    minuend in
+                    minuend - minimumEdge
+                })
+            }
+        }
+        
+        func addBackEdges () {
+            residualNetwork.edges.map {
+                $0.nodes
+                }.filter {
+                    let weight = residualNetwork.weight($0)!
+                    return weight < 0.001 && weight > -0.001
```

Comment:
Would it make sense to have `epsilon` something that could be injected as a parameter?

Reply:
And perhaps we could pull this out of `PerformanceTesting` (and refine it if necessary): [FloatingPoint.isApproximatelyEqual(to:within:)](https://github.com/dn-m/PerformanceTesting/blob/master/Sources/PerformanceTesting/DataSet.swift#L113-L116)

Reply:
Definitely. I get anxious about equating floats even if it is a comparison against 0.0, which I guess should be an exact value in most implementations (could be off about that!).

Reply:
Fair enough!

Reply:
It might be worth it to do some tests around that. I feel like it _should_ be safe to compare against `0.0`.